components/mediation/mediators/dataservices-mediator/org.wso2.micro.integrator.mediator.dataservice/src/main/java/org/wso2/micro/integrator/mediator/dataservice/DataServiceCallMediator.java

```diff
@@ -0,0 +1,390 @@
+            }
+
+        } catch (AxisFault axisFault) {
+            if (synLog.isTraceOrDebugEnabled())
+            {
+                synLog.traceOrDebug(axisFault.getMessage());
+            }
+            synLog.error(axisFault.getMessage());
+            handleException("AxisFault occurred.",axisFault,messageContext);
+        }
+        return true;
+    }
+    
+    private OMElement addRootOperation (org.apache.axis2.context.MessageContext axis2MessageContext,
+                                   MessageContext messageContext) {
+        Operations rootOperations = getOperations();
+        String rootOpName;
+        switch (rootOperations.getType()) {
+            case REQUEST_BOX: {
+                rootOpName = DataServiceCallMediatorConstants.REQUEST_BOX;
+                break;
+            }
+            case BATCH_REQ: {
+                Operation firstOp = (Operation) rootOperations.getOperations().get(0);
+                rootOpName = firstOp.getOperationName() + DataServiceCallMediatorConstants.BATCH_REQ_SUFFIX;
+                break;
+            }
+            case SINGLE_REQ:
+            default: {
+                Operation singleOp = (Operation) rootOperations.getOperations().get(0);
+                rootOpName = singleOp.getOperationName();
+                break;
+            }
+        }
+        QName rootOpQName = new QName(rootOpName);
+        axis2MessageContext.getAxisMessage().getAxisOperation().setName(rootOpQName);
+        OMElement payload = fac.createOMElement(rootOpName, omNamespace);
+        addOperations(rootOperations, payload, messageContext);
+        return payload;
+    }
+
+    private void dispatchToService(org.apache.axis2.context.MessageContext axis2MessageContext,
+                                   MessageContext messageContext) {
+
+        String targetTypeName = getTargetType();
+        String propertyName = getPropertyName();
+        try {
+
+            OMElement omElement = DataServiceProcessor.dispatch(axis2MessageContext);
+
+            if (omElement != null) {
+                if (synLog.isTraceOrDebugEnabled()) {
+                    synLog.traceOrDebug("The result OMElement from the dataservice : " + omElement);
+                }
+
+                //set the result payload as property according to the target type
+                if (DataServiceCallMediatorConstants.TARGET_TYPE.equals(targetTypeName)) {
+                    messageContext.setProperty(propertyName, omElement);
+                    if (synLog.isTraceOrDebugEnabled()) {
+                        synLog.traceOrDebug("The result property : " + messageContext.
+                                getProperty(propertyName));
+                    }
+                } else {
+                    //	set the result payload as envelope in to message context according to the target type
+                    messageContext.getEnvelope().getBody().addChild(omElement);
```

Comment:
do we need to remove NO_ENTITY_BODY while setting payload ?

Reply:
also seems there are some issue in setting the payload back. this gives back the input request even after this.
we can try to set the envelope to axis2MsgCtx. 
```java
    public static void setPayloadInEnvelope(MessageContext axis2MsgCtx, OMElement payload)
            throws ContentBuilderException {

        JsonUtil.removeJsonPayload(axis2MsgCtx);
        try {
            axis2MsgCtx.setEnvelope(TransportUtils.createSOAPEnvelope(payload));
        } catch (AxisFault e) {
            throw new ContentBuilderException("Failed to set XML content.", e);
        }
    }
```

Reply:
Added